src/Illuminate/Routing/UrlGenerator.php

```diff
@@ -393,9 +393,18 @@ public function hasCorrectSignature(Request $request, $absolute = true, array $i
 
         $url = $absolute ? $request->url() : '/'.$request->path();
 
-        $original = rtrim($url.'?'.Arr::query(
-            Arr::except($request->query(), $ignoreQuery)
-        ), '?');
+        $queryString = $request->server->get('QUERY_STRING');
+        foreach ($ignoreQuery as $ignore) {
+            if (strlen($ignore) === 0) {
+                continue;
+            }
+
+            $queryString = ltrim(
+                preg_replace('/(^|&)'.preg_quote($ignore).'((=[^&]*)|(=?(&|$)))/', '', $queryString),
```

Comment:
If anyone wonders about the regex – the equal sign is optional (parameters can be key-only), but when it is, the next character must be either `&` (starting a new parameter) or the end of the string.

Reply:
@Krisell Nice job thinking of the different use cases. Having a chance to think about this some more, I'm starting to feel uncomfortable with the regex approach. At least for me, not sure how others feel. The regex is getting complicated because it has to support more and more use cases. I'm afraid there might be others we haven't thought of. Also it's not readily understandable to someone who might be reading this in the future. Would it work if we exploded the querystring and dealt with each parameter individually?

Reply:
I think you are right @jelib3an, and I think the reason we didn't consider splitting on `&` was that originally signature was the only parameter to exclude.

The change below passes all tests, is easier to understand and also feels more robust. Do you agree?
![image](https://user-images.githubusercontent.com/25909128/121043242-bb93d300-c7b4-11eb-9ec7-cd4da0ca27f6.png)

Reply:
@Krisell 9.x requires 7.4 indeed: 

https://github.com/laravel/framework/blob/95b1899d1a191a99c64cf0848ecba6d9dd201c76/composer.json#L18

Reply:
Agree, looks much nicer! :+1:

Reply:
Committed now. This indeed feels better and the diff is also much smaller.